### PR TITLE
feat(PPDSC-1981): logical props 1 of 6

### PR DESCRIPTION
**Pull request by MadMango**
PPDSC-1981

**What**

1. Background - why this is needed
To introduce layout props to components
2. What did you do
Added layout props to:
- Block
- TextBlock
- Stack
- Button
- Tabs
- AssistiveText
- Label
- Link
- Select (partially, I've done button and panel, not options, there is a separate ticket for Select but I started with this component in the spike so I kept the code)
3. What the reviewers should expect
There should be no visible changes to existing components

**I have done:**
 - [ ] Written unit tests against changes
 - [ ] Written accessibility tests against the component
 - [ ] Written functional tests against the component and/or NewsKit site
 - [ ] Updated relevant documentation
 - [ ] Updated the release notes for the [next release version](https://nidigitalsolutions.jira.com/wiki/spaces/NPP/pages/2349695021/Release%2BNotes%2BCode%2B-%2BWeb%2BDocumentation)
 - [ ] Discussed with QA and/or design about any visual changes

**I have tested manually:**
 - [ ] Renders correctly on IE11
 - [ ] The feature's functionality is working as expected on other browsers such as Safari, Firefox and IE11
 - [ ] There are no accessibility errors reported by WAVE
 - [ ] The screen reader reads the expected text for the tabbable elements.
 - [ ] There are no new errors in the browser console coming from this PR.
 - [ ] When visual test is not added, it renders correctly on different browsers and mobile viewports (Safari, Firefox, small mobile viewport, tablet)
 - [ ] The Playground feature is working as expected

**Before**

**After:**

**Who should review this PR**

**How to test**

### CODE REVIEW

**Comment by pdimova** [comment]
Latest updates: 

1. I've reverted the Tabs changes - this component will be handled in a separate ticket possibly once its building block components get the new changes too.

2. Padding-related props (`paddingInline` or `paddingBlock` for example) can not take squish, inset, and stretch spacing tokens. This currently affects the following components:

Block (spaceInset)
Flag (spaceInset)
Link (new)
SelectButton (spaceInset) 
SelectPanel (spaceInset)

3. We add padding props to all components, but in some cases, such space is not there by design. Such an example is the Link component. This should be discussed with designers

4. Some `spaceInline` props can not be replaced with logical props (and respectively deprecated) as they serve as a gap between internal items (an example is the Stack's spaceInset and spaceStack which sets different spacing than the newly added logical props)


Open to further discuss this in a huddle if needed

File: src/flag/types.ts [hunk 1]
```diff
@@ -7,12 +8,15 @@ export enum FlagSize {
...
+export interface BaseFlagOverrides extends LogicalProps {
   typographyPreset?: MQ<string>;
   stylePreset?: MQ<string>;
   transitionPreset?: TransitionToken | TransitionToken[];
+  /**
+   * @deprecated This property is deprecated and will be removed in the next major release. Use `paddingBlock` and `paddingInline` instead.
+   */
   spaceInset?: MQ<string>;
-  spaceInline?: MQ<string>;
+  spaceInline?: MQ<string>; // Note: LogicalProps can not replace spaceInline! spaceInline = Stack's gap
```
Comment (Vanals) [inline src/flag/types.ts:19]: + 1 for commenting code 🙏

File: src/utils/logical-properties.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  paddingBlockEnd?: MQ<string>;
+  paddingBlock?: MQ<string>;
+}
+
+export interface LogicalProps extends LogicalMargins, LogicalPadding {}
+
+const getResponsiveSpace = (
+  cssProperty: string,
+  props: ThemeProp & {overrides?: unknown},
+  defaultsPath?: string,
```
Comment (Vanals) [inline src/utils/logical-properties.ts:30]: DefaultsPath is needed only if what we are looking for is in the theme?

Comment (pdimova) [reply src/utils/logical-properties.ts:30]: `defaultsPath` takes values for the logical props in the defaults and is an optional argument. 
I doubt there will be defaults for the margin values, but a few components will need default padding ones

**Review by mutebg — DISMISSED** [review]
Well done @pdimova 👏

**Review by Vanals — DISMISSED** [review]

File: src/utils/logical-properties.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+const logicalPadding = (props: ThemeProp, defaultsPath?: string): CSSObject =>
+  generateLogicalProps('padding', defaultsPath)(props);
+
+export const logicalProps = (defaultsPath?: string) => (
+  props: ThemeProp,
+): CSSObject => {
+  const margin = logicalMargins(props, defaultsPath);
+  const padding = logicalPadding(props, defaultsPath);
+  return deepMerge(margin, padding);
```
Comment (Xin00163) [inline src/utils/logical-properties.ts:104]: I wondered if this is supposed to return values for `block` & `label`. I tried `console.log(deepMerge(margin, padding))` and it's returning an empty {} for me.

File: src/link/internal-link.tsx [hunk 4]
```diff
@@ -9,12 +9,14 @@ import {useTheme} from '../theme';
...
 import {useHasMounted} from '../utils/hooks';
 import {TextBlock} from '../text-block';
+import {logicalProps} from '../utils/logical-properties';
 import {getTransitionPreset} from '../utils/style/transition-preset';
 
 const StyledLink = styled.a<InternalLinkProps>`
   display: inline-block;
   ${getTransitionPreset(`link`, '')};
   ${getStylePreset('link', '')}
+  ${logicalProps('link')}
```
[the remark this answers is not in the document]
Comment (pdimova) [reply src/link/internal-link.tsx:19]: Not all components currently have defaults, even some of them won't have such defaults at all in my opinion but you're right that we break the consistency if just a few calls to the `logicalProps` take `defaultProps` as an argument.
So I've added default paths everywhere possible

File: src/utils/logical-properties.ts (hunk 2 above)
Comment (pdimova) [reply src/utils/logical-properties.ts:30]: you can check the comment bellow but I've added the default paths everywhere possible

**Review by mutebg — APPROVED** [review]

**Review by Vanals — APPROVED** [review]

**Pull request merged**